tests/functional/pyocf/rio.py

```diff
@@ -0,0 +1,291 @@
+#
+# Copyright(c) 2021 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+#
+
+from ctypes import c_int, c_void_p, CFUNCTYPE
+from enum import Enum, auto
+from random import shuffle, choice
+from dataclasses import dataclass
+from datetime import timedelta, datetime
+from itertools import cycle
+from threading import Thread, Condition, Event
+
+from pyocf.utils import Size
+from pyocf.types.volume import Volume
+from pyocf.types.io import Io, IoDir
+from pyocf.types.data import Data
+
+
+class ReadWrite(Enum):
+    READ = auto()
+    WRITE = auto()
+    RANDREAD = auto()
+    RANDWRITE = auto()
+    RANDRW = auto()
+
+    def is_random(self):
+        return self in [self.RANDREAD, self.RANDWRITE, self.RANDRW]
+
+
+class IoGen:
+    def __init__(self, seed, extent, blocksize=512, random=False, randommap=True):
+        self.seed = seed
+        self.random = random
+        self.randommap = randommap
+
+        gen = list(range(extent[0], extent[1], blocksize))
+        self.cycle_len = len(gen)
+
+        if random:
+            shuffle(gen)
+
+        if self.randommap:
+            self.gen = cycle(gen)
+        else:
+            self.gen = gen
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.random and not self.randommap:
+            return choice(self.gen)
+
+        return next(self.gen)
+
+
+@dataclass
+class JobSpec:
+    readwrite: ReadWrite = ReadWrite.READ
+    randseed: int = 1
+    rwmixwrite: int = 50
+    randommap: bool = True
+    bs: Size = Size.from_B(512)
+    offset: Size = Size(0)
+    njobs: int = 1
+    qd: int = 1
+    size: Size = Size(0)
+    io_size: Size = Size(0)
+    target: Volume = None
+    time_based: bool = False
+    time: timedelta = None
+    continue_on_error: bool = False
+
+    def merge(self, other):
+        # TODO implement
+        return other
+
+
+class Rio:
+    class RioThread:
+        pass
+
+
+class Rio:
+    class RioThread(Thread):
```

Comment:
Does it have any purpose to define this class twice or is it just an oversight?

Reply:
Whoops, that wasn't supposed to be here